zipkin-ui/js/skew.js

```diff
@@ -63,105 +66,170 @@ class Node {
   }
 
   toString() {
-    if (this._value) return `Node(${JSON.stringify(this._value)})`;
-    return 'Node()';
+    if (this._span) return `SpanNode(${JSON.stringify(this._span)})`;
+    return 'SpanNode()';
   }
 }
 
-/*
- * Some operations do not require the entire span object. This creates a tree given (parent id,
- * id) pairs.
- */
-class TreeBuilder {
+// In javascript, dict keys can't be objects
+function keyString(id, shared = false, endpoint) {
+  const sharedV = shared === true;
+  const endpointString = endpoint ? JSON.stringify(endpoint) : 'x';
+  return `${id}-${sharedV}-${endpointString}`;
+}
+
+class SpanNodeBuilder {
   constructor(params) {
-    const {traceId, debug = false} = params;
-    this._mergeFunction = (existing, update) => existing || update; // first non null
-    if (!traceId) throw new Error('traceId was undefined');
-    this._traceId = traceId;
+    const {debug = false} = params;
     this._debug = debug;
-    this._rootId = undefined;
-    this._rootNode = undefined;
-    this._entries = [];
-      // Nodes representing the trace tree
-    this._idToNode = {};
-      // Collect the parent-child relationships between all spans.
-    this._idToParent = {};
+    this._rootSpan = undefined;
+    this._keyToNode = {};
+    this._spanToParent = {};
   }
 
-  // Returns false after logging on debug if the value couldn't be added
-  addNode(parentId, id, value) {
-    if (parentId && parentId === id) {
-      if (this._debug) {
-        /* eslint-disable no-console */
-        console.log(`skipping circular dependency: traceId=${this._traceId}, spanId=${id}`);
-      }
-      return false;
+  /*
+   * We index spans by (id, shared, localEndpoint) before processing them. This latter fields
+   * (shared, endpoint) are important because in zipkin (specifically B3), a server can share
+   * (re-use) the same ID as its client. This impacts processing quite a bit when multiple servers
+   * share one span ID.
+   *
+   * In a Zipkin trace, a parent (client) and child (server) can share the same ID if in an
+   * RPC. If two different servers respond to the same client, the only way for us to tell which
+   * is which is by endpoint. Our goal is to retain full paths across multiple endpoints. Even
+   * though instrumentation should be configured in such a way that a client never sends the same
+   * span ID to multiple servers, it can happen. Accordingly, we index defensively including any
+   * endpoint data that might be available.
+   */
+  _index(span) {
+    let idKey;
+    let parentKey;
+    const shared = span.shared === true; // guards against undefined
+
+    if (shared) {
+      // we need to classify a shared span by its endpoint in case multiple servers respond to the
+      // same ID sent by the client.
+      idKey = keyString(span.id, true, span.localEndpoint);
```

Comment:
I think `span.localEndpoint` is a little large for using as key string...
I think that it is better to use a hash value... 🤔

Reply:
maybe.. doesn't seem obvious for a hashCode replacement. do you know one or should I manually write one like we have in java?